File: kernel/drivers/serial.c
/*
 serial.c
 As part of the Avery project
 Created by Maxims Enterprise in 2025
 --------------------------------------------------
 Description: Serial COM1 bindings for writing to serial
 Copyright (c) 2025 Maxims Enterprise
*/

#include "drivers/serial.h"
#include "common.h"

void serial_init() {
    outb(SERIAL_PORT_COM1 + 1, 0x00);
    outb(SERIAL_PORT_COM1 + 3, 0x80);
    outb(SERIAL_PORT_COM1 + 0, 0x03);
    outb(SERIAL_PORT_COM1 + 1, 0x00);
    outb(SERIAL_PORT_COM1 + 3, 0x03);
    outb(SERIAL_PORT_COM1 + 2, 0xC7);
    outb(SERIAL_PORT_COM1 + 4, 0x0B);
}

u32 is_serial_read() { return inb(SERIAL_PORT_COM1 + 5) & 0x20; }

void serial_write(char data) {
    while (is_serial_read() == 0)
        ;
    outb(SERIAL_PORT_COM1, data);
}

void serial_write_str(char *data) {
    while (*data != 0) {
        serial_write(*data);
        data++;
    }
}

File: kernel/drivers/serial/serial.zig
const mem = @import("memory");
const sys = @import("system");
const str = @import("string");

const SERIAL_PORT: u16 = 0x3F8;

pub fn initSerial() void {
    sys.outb(SERIAL_PORT + 1, 0x00);
    sys.outb(SERIAL_PORT + 3, 0x80);
    sys.outb(SERIAL_PORT + 0, 0x03);
    sys.outb(SERIAL_PORT + 1, 0x00);
    sys.outb(SERIAL_PORT + 3, 0x03);
    sys.outb(SERIAL_PORT + 2, 0xC7);
    sys.outb(SERIAL_PORT + 4, 0x0B);
    sys.outb(SERIAL_PORT + 5, 0x01);
}

fn isTransmitReady() bool {
    return (sys.inb(SERIAL_PORT + 5) & 0x20) != 0;
}

pub fn writeChar(c: str.char) void {
    while (!isTransmitReady()) {}
    sys.outb(SERIAL_PORT, c);
}

pub fn writeString(s: str.String) void {
    for (s.iterate()) |c| {
        if (c == '\n') writeChar('\r');
        writeChar(c);
    }
}
